nexus/src/app/allowed_source_ips.rs

```diff
@@ -0,0 +1,147 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2024 Oxide Computer Company
+
+//! Nexus methods for operating on source IP allowlists.
+
+use std::net::IpAddr;
+
+use nexus_db_queries::context::OpContext;
+use nexus_types::external_api::params;
+use nexus_types::external_api::views::AllowedSourceIps;
+use omicron_common::api::external;
+use omicron_common::api::external::Error;
+
+impl super::Nexus {
+    /// Fetch the allowlist of source IPs that can reach user-facing services.
+    pub async fn allowed_source_ips_view(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<AllowedSourceIps, Error> {
+        self.db_datastore
+            .allowed_source_ips_view(opctx)
+            .await
+            .map(AllowedSourceIps::from)
+    }
+
+    /// Upsert the allowlist of source IPs that can reach user-facing services.
+    pub async fn allowed_source_ips_upsert(
+        &self,
+        opctx: &OpContext,
+        remote_addr: IpAddr,
+        params: params::AllowedSourceIpsUpdate,
+    ) -> Result<(), Error> {
+        if let external::AllowedSourceIps::List(list) = &params.allowed_ips {
+            // Size limits on the allowlist.
+            const MAX_ALLOWLIST_LENGTH: usize = 1000;
+            if list.len() > MAX_ALLOWLIST_LENGTH {
+                let message = format!(
+                    "Source IP allowlist is limited to {} entries, found {}",
+                    MAX_ALLOWLIST_LENGTH,
+                    list.len(),
+                );
+                return Err(Error::invalid_request(message));
+            }
+            if list.is_empty() {
+                return Err(Error::invalid_request(
+                    "Source IP allow list must not be empty",
+                ));
+            }
+
+            // Some basic sanity-checks on the addresses in the allowlist.
+            //
+            // The most important part here is checking that the source address
+            // the request came from is on the allowlist. This is our only real
+            // guardrail to prevent accidentally preventing any future access to
+            // the rack!
+            for entry in list.iter() {
+                if !entry.contains(remote_addr) {
```

Comment:
This requires every entry in `list` to contain `remote_addr` - we only care that at least one entry does so, right?

Reply:
D'oh, yep great catch.

Reply:
Done in d72f7dc